### PR TITLE
Firestore emulator live update on rules file changes

### PR DESCRIPTION
**Pull request by samtstern**
Watches the file specified in `firebase.json` for changes and updates the rules in the emulator.

**Successful Change**
Screenshot from 2019-06-28 08-34-14

**Unsuccessful Change**
image

I styled the error printing based on what the TypeScript compiler does in "watch" mode.

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
What is the behavior you'd expect if you have a syntax error in your rules?

**Comment by samtstern** [comment]
@ryanpbrewster I would expect the emulator to reject the ruleset with a 4xx error ... what does it actually do?

**Comment by ryanpbrewster** [comment]
It does indeed fail, and as a result does not update the rules. You end up in a situation where you're editing your rules and the edits aren't taking effect. Unless you look at the emulator suite logging (often in the background or in some other terminal) you won't be able to figure out why.

**Comment by samtstern** [comment]
@ryanpbrewster here's what the logs look like now with some successful/unsuccessful updates in there:
```
$ firebase emulators:start --only firestore
i  Starting emulators: ["firestore"]
i  firestore: Logging to firestore-debug.log
✔  firestore: Emulator started at http://localhost:8080
i  firestore: For testing set FIRESTORE_EMULATOR_HOST=localhost:8080
i  firestore: Change detected, updating rules...
✔  firestore: Rules updated.
i  firestore: Change detected, updating rules...
⚠  Failed to update rules.
⚠  Error compiling rules:
L4:26 Missing 'if' before conditional expression.
L4:29 Unexpected 'true'.
i  firestore: Change detected, updating rules...
✔  firestore: Rules updated.
```

Honestly I think that's pretty nice.

**Comment by samtstern** [comment]
To be clear on next steps, @ryanpbrewster is going to modify `setRules` as follows:

  * The endpoint will take some sort of `ignore` parameter that allows the set to succeed ignoring certain classes of rules issues.  We will use `ignore=ERROR` to allow the set to always succeed.
  * The endpoint will return a structured proto describing rules issues so that the client can print them / act on them.
  * If invalid rules are set, the emulator will treat them like `read, write: if false;` and deny all requests until valid rules are set.

File: src/emulator/controller.ts [hunk 1]
```diff
@@ -102,6 +103,8 @@ export async function startAll(options: any): Promise<void> {
   const targets: string[] = filterTargets(options, VALID_EMULATOR_STRINGS);
   options.targets = targets;
 
+  const projectId = getProjectId(options, false);
```
Comment (samtstern) [inline src/emulator/controller.ts]: What happens when I don't have `firebase.json`?

File: package-lock.json [hunk 2]
```diff
@@ -230,6 +230,7 @@
           "resolved": "https://registry.npmjs.org/gaxios/-/gaxios-1.8.4.tgz",
           "integrity": "sha512-BoENMnu1Gav18HcpV9IleMPZ9exM+AvUjrAOV4Mzs/vfz2Lu/ABv451iEXByKiMPn2M140uul1txXCg83sAENw==",
           "dev": true,
+          "optional": true,
```
Comment (samtstern) [inline package-lock.json]: Gotta get rid of all this nonsense.

Comment (samtstern) [reply package-lock.json]: @yuchenshi I can't tell if I am wrong or `master` is wrong, I am using `npm` 6.9.2

Comment (yuchenshi) [reply package-lock.json]: I think master is wrong since https://github.com/firebase/firebase-tools/pull/1433/files

Comment (samtstern) [reply package-lock.json]: Ah thank you!

**Comment by samtstern** [comment]
The Travis build passed but I think the webhook back to GitHub must have been dropped:
https://travis-ci.org/firebase/firebase-tools/builds/553869724?utm_source=github_status&utm_medium=notification

So I am going to merge this.

**Pull request merged**